File: go/array/leetcode/move_zero.go
package main

//移动零：https://leetcode.com/problems/move-zeroes/

func moveZeroes1(nums []int) {
	i := 0
	for j := 0; j < len(nums); j++ {
		if nums[j] != 0 {
			if i != j {
				nums[i], nums[j] = nums[j], nums[i]
			}
			i++
		}
	}
}

func moveZeroes2(nums []int) {
	i := 0
	for j := 0; j < len(nums); j++ {
		if nums[j] != 0 {
			if i != j {
				nums[i] = nums[j]
				nums[j] = 0
			}
			i++
		}
	}
}

func moveZeros3(nums []int) {
	i, j := 0, 0
	for j < len(nums) {
		if nums[j] != 0 {
			nums[i], nums[j] = nums[j], nums[i]
			i++
		}
		j++
	}
}

File: python/array/leetcode/move_zero.py
from typing import List


# 移动零


class Solution:

    # 新开一个数组
    def moveZeroes1(self, nums: List[int]) -> None:
        temp, k = [0] * len(nums), 0
        for n in nums:
            if n != 0:
                temp[k] = n
                k += 1
        nums[:] = temp[:]

    # 双指针解法
    def moveZeroes2(self, nums: List[int]) -> None:
        k = 0
        for i, v in enumerate(nums):
            if v != 0:
                nums[i], nums[k] = nums[k], nums[i]
                k += 1
